core/src/cluster_slots.rs

```diff
@@ -77,18 +77,14 @@ impl ClusterSlots {
                     .zip(std::iter::repeat((epoch_slots.from, stake)))
             })
             .into_group_map();
-        let slot_nodes_stakes: Vec<_> = {
```

Comment:
This is a local variable which will be freed imminently.
How does the new code take less memory than the current code?

The code was written this way to minimize the amount of time we hold a write lock on `self.cluster_slots`.

Reply:
> When you collected the memory profile, did you check if the node has caught up and is in sync with the cluster? or was it falling behind?
> 
> I am suspecting it was falling further behind, which causes cluster-slots to consume more memory.
Possible. With profiling, the validator runs slower. As a result, it may falling behind.

Reply:
> > When you collected the memory profile, did you check if the node has caught up and is in sync with the cluster? or was it falling behind?
> > I am suspecting it was falling further behind, which causes cluster-slots to consume more memory.
> > Possible. With profiling, the validator runs slower. As a result, it may falling behind.

I checked the log. yes. it is behind.

Reply:
you can monitor node status using `solana-validator monitor` or `solana catchup` tools.

Reply:
![image](https://user-images.githubusercontent.com/219428/166806464-89d43539-958b-493a-9bae-7890b0a58972.png)
The cluster_slots are around 5.2K. Sound like it is expensive to collect these 5.2K vector and iterate them. 
I will add some measurement to compare the performance of the new code.

Reply:
The following shows two runs, one with the old and the other with new code. The first number is the *cluster_slots.len()* and the second number is the time in micros. The new code is about 37% faster. 

old 
```
[2022-05-04T19:03:19.450941125Z INFO  solana_core::cluster_slots] haha cluster_slot 4676 2503078
[2022-05-04T19:03:24.015446946Z INFO  solana_core::cluster_slots] haha cluster_slot 4675 2385321
[2022-05-04T19:03:29.059829276Z INFO  solana_core::cluster_slots] haha cluster_slot 4677 2317881
```

new
```
[2022-05-04T19:58:05.780770207Z INFO  solana_core::cluster_slots] haha cluster_slot 5217 1966092
[2022-05-04T19:58:09.970466150Z INFO  solana_core::cluster_slots] haha cluster_slot 5219 1940792
[2022-05-04T19:58:14.039771855Z INFO  solana_core::cluster_slots] haha cluster_slot 5221 1870872
```


l

Reply:
As I mentioned, the code was written this way to minimize the amount of time we hold a write lock on `self.cluster_slots`.

Reply:
The total time to process cluster slot update is 4.2s. And about half of the time is spent on update cluster slot. 
```
[2022-05-04T19:58:32.596074602Z INFO  solana_core::cluster_slots] haha cluster_slot 5224 1940228
[2022-05-04T19:58:32.596243594Z INFO  solana_metrics::metrics] datapoint: cluster_slots_service-timing lowest_slot_elapsed=117i process_cluster_slots_updates_elapsed=4228825i
```

Reply:
> As I mentioned, the code was written this way to minimize the amount of time we hold a write lock on `self.cluster_slots`.

What are the readers for self.cluster_slots. How critical they are? 
I think another question to answer is how much longer do we hold the lock in the new code than the old code. 
I will measure it.

Reply:
> The total time to process cluster slot update is 4.2s.

I am suspecting that you are quoting again from a node which cannot catch-up with the cluster.
What I see in the dashboards it is more like 200ms on average.
https://github.com/solana-labs/solana/pull/24936#pullrequestreview-961961999

Please do not evaluate code changes based on a node which has fallen behind and cannot catch-up. That can actually make things worse for the nodes which processes in sync with the cluster.

Reply:
So the size of cluster_slots is linear to how far the nodes are behind from the tip of the network?